### PR TITLE
Decompile functions accessing window list

### LINKED ISSUE #169: Segfault on startup/window resize event
**OS:**  Arch Linux
**Version:**  18.02
**Commit/Build:**  68bd1ac2c118124f322ce9d4ca70be1f91f4b36d

When I start OpenLoco on my system, my window manager (i3) resizes the window. Because this happens immediately after the window is created it is one of the first events that is handeled in `ui::process_messages`. The window structure is not yet initialized with correct values and a segfault occurs in gui.cpp line 212, because `window->widgets[0]` is NULL.

**Steps to reproduce:**
1. Start OpenLoco while using the i3 window manager

**Is the issue reproducible in Locomotion (vanilla)?**
No, no Linux version

**Screenshots / Video:**
openlocosegfault

### PR DESCRIPTION
**Pull request by marijnvdwerf**
Fixes #169

### CODE REVIEW

File: src/openloco/interop/hooks.cpp [hunk 1]
```diff
@@ -394,7 +394,7 @@ static int32_t STDCALL lib_CreateFileA(
     {
         pFILE = fopen(lpFileName, "r");
     }
-    else if (dwDesiredAccess == GENERIC_WRITE && dwCreationDisposition == CREATE_ALWAYS)
+    else if (dwDesiredAccess == GE2170NERIC_WRITE && dwCreationDisposition == CREATE_ALWAYS)
```
Comment (AaronVanGeffen) [inline src/openloco/interop/hooks.cpp]: Accidental change?

File: src/openloco/windowmgr.cpp [hunk 2]
```diff
@@ -600,4 +651,146 @@ namespace openloco::ui::windowmgr
...
+    {
+        window* w = _windows_end;
+        while (w >= _windows)
+        {
+            w->call_21();
+        }
+    }
+
+    // 0x004CD296
+    void relocate_windows()
```
Comment (AaronVanGeffen) [inline src/openloco/windowmgr.cpp]: Is this cascading the windows?

File: src/openloco/windowmgr.cpp [hunk 3]
```diff
@@ -14,6 +14,45 @@ namespace openloco::ui::windowmgr
         constexpr uint16_t by_type = 1 << 7;
     }
 
+    class Container
```
Comment (AaronVanGeffen) [inline src/openloco/windowmgr.cpp]: I'm not sure I see the need for such a class. Can we make do with a `std::vector<window*> foo(_windows, _windows_end);` instead?

Comment (IntelOrca) [reply src/openloco/windowmgr.cpp]: @AaronVanGeffen You can't use `std::vector<window*>`, but you could do `std::array<window>*` pointing to the address of where the windows are.

**Comment by marijnvdwerf** [comment]
As far as I'm aware, this is the current state of implementation+hooking: 

image
green is in this PR, yellow is (in some form) in #174

**Comment by marijnvdwerf** [comment]
@AaronVanGeffen Can you review the functions done so far? The remaining ones are all quite big, and I think it would be better to tackle them in a new PR.

File: src/openloco/windowmgr.cpp [hunk 4]
```diff
@@ -48,85 +258,189 @@ namespace openloco::ui::windowmgr
...
+
+                close(&w);
+                repeat = true;
+                break;
+            }
+        }
     }
 
     // 0x004CC692
     void close(window_type type, uint16_t id)
```
Comment (AaronVanGeffen) [inline src/openloco/windowmgr.cpp]: uint16_t -> window_number

**Comment by AaronVanGeffen** [comment]
## Hooks

| Reviewed | Loco address |
|----|--------|
|    | 4CC6EA |
|    | 4CD3D0 |
|    | 4C5FC8 |
|    | 4C9B56 |
|    | 4CE438 |
|    | 4C9A95 |
|    | 4C9AFA |
|    | 4CB966 |
|    | 4CC692 |
|    | 45F18B |
|    | 4CD296 |
|    | 4CEE0B |
|    | 4B93A5 |
|    | 4BF089 |
|    | 4C6202 |

## Functions
| Reviewed | Loco address | OpenLoco implementation |
|----|--------|--------------------|
|    | 4CE438 | window* get_main() |
| x  | 4C9B56 | window* find(window_type type) |
| x  | 4C9B56 | window* find(window_type type, window_number number) |
|    | 4C9A95 | window* find_at(int16_t x, int16_t y) |
|    | 4C9AFA | window* find_at_alt(int16_t x, int16_t y) |
|    | 4CB966 | void invalidate(window_type type) |
|    | 4CB966 | void invalidate(window_type type, window_number number) |
|    | 4CB966 | void invalidate_widget(window_type type, window_number number, uint8_t widget_index) |
|    | 4CC692 | void close(window_type type) |
|    | 4CC692 | void close(window_type type, uint16_t id) |
|    | 45F18B | sub_45F18B() |
|    | 4CD296 | void relocate_windows() |
|    | 4CEE0B | void sub_4CEE0B(window* self) |
|    | 4B93A5 | sub_4B93A5(uint16_t arg) |
|    | n/a    | window_scroll_wheel_input(ui::window* window, widget_index widgetIndex, int wheel) |
|    | 4C628E | window_wheel_input(window* window, int wheel) |
| x  | 45EFDB | sub_45EFDB(ui::window* window) |
| x  | 45F015 | sub_45F015(ui::window* window) |
| x  | 45F04F | sub_45F04F(ui::window* window) |
| x  | 45F0ED | sub_45F0ED(ui::window* window) |
| x  | 49771C | sub_49771C() |
| x  | 48DDC3 | sub_48DDC3() |
|    | n/a    | all_wheel_input() |

**Pull request merged**